spring-integration-file/src/main/java/org/springframework/integration/file/FileWritingMessageHandler.java

```diff
@@ -608,4 +728,188 @@ else if (destinationDirectoryToUse instanceof File) {
 		return destinationDirectory;
 	}
 
+	private synchronized FileState getFileState(final File fileToWriteTo, boolean isString)
+			throws FileNotFoundException {
+		String absolutePath = fileToWriteTo.getAbsolutePath();
+		FileState state;
+		boolean appendNoFlush = FileExistsMode.APPEND_NO_FLUSH.equals(this.fileExistsMode);
+		if (appendNoFlush) {
+			state = this.fileStates.get(absolutePath);
+			if (state != null && ((isString && state.stream != null) || (!isString && state.writer != null))) {
+				state.close();
+				state = null;
+				this.fileStates.remove(absolutePath);
+			}
+			if (state == null) {
+				if (isString) {
+					state = new FileState(createWriter(fileToWriteTo, true));
+				}
+				else {
+					state = new FileState(createOutputStream(fileToWriteTo, true));
+				}
+				this.fileStates.put(absolutePath, state);
+			}
+			state.lastWrite = Long.MAX_VALUE; // prevent flush while we write
+		}
+		else {
+			state = null;
+		}
+		return state;
+	}
+
+	private BufferedWriter createWriter(final File fileToWriteTo, final boolean append) throws FileNotFoundException {
+		return new BufferedWriter(new OutputStreamWriter(new FileOutputStream(fileToWriteTo, append), this.charset),
+				this.bufferSize);
+	}
+
+	private BufferedOutputStream createOutputStream(File fileToWriteTo, final boolean append)
+			throws FileNotFoundException {
+		return new BufferedOutputStream(new FileOutputStream(fileToWriteTo, append), this.bufferSize);
+	}
+
+	/**
+	 * When using {@link FileExistsMode#APPEND_NO_FLUSH}, you can send a message to this
+	 * method to flush any file(s) that needs it. By default, the payload must be a regular
+	 * expression ({@link String} or {@link Pattern}) that matches the absolutePath
+	 * of any in-process files. However, if a custom {@link FlushPredicate} is provided,
+	 * the payload can be of any type supported by that implementation.
+	 * @since 4.3
+	 */
+	@Override
+	public synchronized void trigger(Message<?> message) {
+		flushIfNeeded(this.flushPredicate, message);
+	}
+
+	/**
+	 * When using {@link FileExistsMode#APPEND_NO_FLUSH} you can invoke this method to
+	 * selectively flush open files. For each open file the supplied
+	 * {@link FlushPredicate#shouldFlush(String, long, Message)}
+	 * method is invoked and if true is returned, the file is flushed.
+	 * @param flushPredicate the {@link FlushPredicate}.
+	 * @since 4.3
+	 */
+	public void flushIfNeeded(FlushPredicate flushPredicate) {
+		flushIfNeeded(flushPredicate, null);
```

Comment:
Yeah... What I don't like here that `null` as a `message` argument for the `shouldFlush`:
```
fileWritingMessageHandler.flushIfNeeded((fileAbsolutePath, lastWrite,  filterMessage) -> ...);
```
That `filterMessage` does not make sense in this context at all.
So, we should consider to introduce one more `FlushPredicate` contract to pass in only the flashable entry.
I mean only with those two first arguments.